patches/server/0738-Fix-entities-glitching-through-blocks-pushed-by-pist.patch

```diff
@@ -0,0 +1,80 @@
+From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
+From: Peter Svensson <petersv@psv.nu>
+Date: Mon, 2 Aug 2021 12:06:05 +0200
+Subject: [PATCH] Fix entities glitching through blocks pushed by pistons
+ (#5189)
+
+Fix #5189 where entities in slime blocks are unreliable.
+
+Add a general mechanism where the entity knows how many ticks into
+the future that it has to be fully ticked. ActivationRange respects
+that and does not cull any of those ticks to the entity.
+
+diff --git a/src/main/java/net/minecraft/world/entity/Entity.java b/src/main/java/net/minecraft/world/entity/Entity.java
+index 896d892237b29eb404398db07264eb6f04786754..7981965bfb554709d3b25eb8ae98f3e6853b9e64 100644
+--- a/src/main/java/net/minecraft/world/entity/Entity.java
++++ b/src/main/java/net/minecraft/world/entity/Entity.java
+@@ -690,6 +690,11 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource, n
+ 
+         this.firstTick = false;
+         this.level.getProfiler().pop();
++        // Paper start - track when all ticks are required for the entity, e.g. after a piston push
++        if (this.isTickRequired()) {
++           --this.remainingExactTicks;
++        }
++        // Paper end
+     }
+ 
+     public void setSharedFlagOnFire(boolean onFire) {
+@@ -3991,5 +3996,16 @@ public abstract class Entity implements Nameable, EntityAccess, CommandSource, n
+     public boolean isTicking() {
+         return ((ServerChunkCache) level.getChunkSource()).isPositionTicking(this);
+     }
++ 
++    // Track when all ticks are required for the entity, e.g. after a piston push
++    private int remainingExactTicks = 0;
++
++    public void requestExactTicks(int ticks) {
++        this.remainingExactTicks = Math.max(ticks, this.remainingExactTicks);
++    }
++
++    public boolean isTickRequired() {
++        return (this.remainingExactTicks > 0);
++    }
+     // Paper end
+ }
+diff --git a/src/main/java/net/minecraft/world/level/block/piston/PistonMovingBlockEntity.java b/src/main/java/net/minecraft/world/level/block/piston/PistonMovingBlockEntity.java
+index 87bedba9ab495edcce289c6665271d92b7165944..04c3ef614b65147f724b92705394b3d9573a3a98 100644
+--- a/src/main/java/net/minecraft/world/level/block/piston/PistonMovingBlockEntity.java
++++ b/src/main/java/net/minecraft/world/level/block/piston/PistonMovingBlockEntity.java
+@@ -140,6 +140,7 @@ public class PistonMovingBlockEntity extends BlockEntity {
+                                 }
+ 
+                                 entity.setDeltaMovement(e, g, h);
++                                entity.requestExactTicks(2); // Paper - track when all ticks are required for the entity, e.g. after a piston push
```

Comment:
Do you want me to switch from tracking/counting down) a certain number of ticks to instead track the last tick to be immune in absolute terms?

The patch originally started with 10 ticks, but Proimyst asked it to be reduced. 2 ticks are sufficient for entities stuck in pistons, as they will complete the movment in 2 ticks from reading the code and testing. The original 10 ticks were to allow for entities being bounced (not stuck inside slime) to complete ballisitc movement. After some back and forth we decided to drop that as we had no test cases for how long was sufficient anyway.  Do you want me to up the immunity again to 20 ticks (one second)?

Whis is the semantic for the this.activatedTick?

Reply:
If you are 100% certain 2 ticks are enough, then that's fine. In my testing it wasn't but I had some other code changes flying around as well, so idk. activatedTick is basically exactly what you are doing without the force part, since it additionally does that skip every 4 ticks if not immune. so if activatedTick >= currentTick, the entity should be ticked (so you do the same as that, but checked in the immunity method).

Reply:
I may not follow exactly here. Did you mean to both move activatedTime approriately into the future + also having an entity-centric immunization as in my patch but with updated semantics?

Reply:
No to the first. But looking at the activatedTick usages, you don't necessarily have to add the codeline I provided. Just keep the number at whatever you are sure works and only update the immunity `entity.activationImmunityTick = MinecraftServer.currentTick + <ticks it should be active for>;`

Reply:
If activatedTick is behind the currentTick, the entity is marked as inactive and doesn't tick at all unless it is immune given by the immune check in ActivationRange. Sidenote: If *this* particular spot didn't exist and you only had to account for the other place in the move() method, you could even just use `entity.pistonDeltasGameTime` and shove it into the immunity method.. but you can't, since this is here

So the correct way of keeping it active is by "immunizing" it. But since the immunity check method is used nowhere else, it doesn't matter wheter your new check is put into that method or exactly where you have it at the moment (tho you *do* have to set activatedTick here *if* you moved it into the immune check method).

Reply:
A little more digging on my part: Today the activation of entities moved by being stuck on honey or bounced _on_ slime blocks (and some other cases) get activated fo 20 ticks in Entity.move() due to the movementType being MoverType.PISTON. For items that are stuck _in_ slime instead get moved by Entity.setDeltaMovement that keeps being updated by the PistonMovingBlockEntity. 

In my patch I did the immunization for both cases in PistonMovingBlockEntity. However, maybe the immunzation for the non-stuck-in-slime should be moved next to the activatedTime update in Entity.move() ?

What concerns me is that I cannot find where the entities stuck inside slime received their activation before my patch. They were clearly activated as the failure rate was 1/4. Was it due to short distance to the player? I propose to add setting both activatedTick and actiavtedImmunityTick when doing the setDeltaMovememnt call and moving the test into the checkEntityimmunities() as you suggested.

Reply:
Yep, that should work. 👍

Reply:
Is it ok to blindly assign activatedTick / activatedimmunityTick with currentTick+xxxx, or sohuld I use max(activatedtick, currentTick+xx) ?

Reply:
Good point. The piston type in the move method currently is the only one that would theoretically not recover itself, but better save then sorry

Reply:
MinecraftServer does not seem to be used in PistonMovingBlockEntity at the moment. What is the best way to get at the currentTick there?

Reply:
I believe it is a static value, so using a fully qualified import, `net.minecraft.server.MinecraftServer.currentTick`